File: sim/src/physical/render.ts
import { PANEL_W as W, PANEL_H, TUBE_HEIGHT_MAX as MAX_H } from '@spec/layout';
import { PHYSICAL_META, validatePhysicalParams, type PhysicalParams } from './model';
import { physicalLayout, traceRow } from './geometry';
import { beer, type RGB } from './optics';
import { backingLight } from './lighting';
import { markAt } from './marks';
import { blendLinear565, quantize } from './output';
import { decode } from './optics';
import { DEFAULT_PHYSICAL_APPEARANCE, type PhysicalAppearance } from './appearance';

/** Same fixed maximum layer capacity as firmware; no material-dependent pool growth. */
export class PhysicalRenderer {
  readonly fb = new Uint16Array(W * PANEL_H);
  readonly layers = new Uint16Array(2 * 2 * W * MAX_H);
  private key = '';
  lastRebuildMs = 0;
  private rebuild(p: PhysicalParams, appearance: PhysicalAppearance): void {
    const { H } = physicalLayout(p), R = p.innerRadiusMm + p.wallThicknessMm;
    const irradiance = backingLight(p);
    const srgb = (hex: string): RGB => {
      const value = Number.parseInt(hex.slice(1), 16);
      return {
        r: decode(((value >> 16) & 255) / 255),
        g: decode(((value >> 8) & 255) / 255),
        b: decode((value & 255) / 255),
      };
    };
    const background = srgb(appearance.background), digits = srgb(appearance.digits);
    for (let medium = 0; medium < 2; medium++) for (let y = 0; y < H; y++) {
      const path = traceRow(((y + 0.5) / H * 2 - 1) * R, medium === 1, p);
      const gain = path.transmission * irradiance;
      const colour = (albedo: number, tint: RGB): RGB => ({
        r: p.exposure * (path.reflection + gain * albedo * tint.r * beer(p.absorptionR, path.distance)),
        g: p.exposure * (path.reflection + gain * albedo * tint.g * beer(p.absorptionG, path.distance)),
        b: p.exposure * (path.reflection + gain * albedo * tint.b * beer(p.absorptionB, path.distance)),
      });
      const bg = quantize(colour(p.backingReflectance, background));
      const mark = quantize(colour(p.marksReflectance, digits));
      const sy = (path.backY / R + 1) * H * 0.5;
      for (let tube = 0; tube < 2; tube++) {
        const offset = (tube * 2 + medium) * W * MAX_H + y * W;
        for (let x = 0; x < W; x++) this.layers[offset + x] = path.transmission > 0 && markAt(x, sy, H, tube) ? mark : bg;
      }
    }
  }
  render(p: PhysicalParams, hoursFill: number, minutesFill: number,
         appearance: PhysicalAppearance = DEFAULT_PHYSICAL_APPEARANCE): Uint16Array {
    const key = PHYSICAL_META.map(f => p[f.key]).join(',') + `|${appearance.background}|${appearance.digits}`;
    this.lastRebuildMs = 0;
    if (key !== this.key) {
      validatePhysicalParams(p);
      const t = performance.now(); this.rebuild(p, appearance); this.lastRebuildMs = performance.now() - t; this.key = key;
    }
    const { H, yH, yM } = physicalLayout(p);
    this.fb.fill(0);
    for (let tube = 0; tube < 2; tube++) {
      const fill = tube ? minutesFill : hoursFill;
      if (!Number.isFinite(fill)) throw new Error('Non-finite physical fill');
      const edge = Math.max(0, Math.min(1, fill)) * W, xi = Math.floor(edge), alpha = edge - xi;
      const y0 = tube ? yM : yH, dry = tube * 2 * W * MAX_H, wet = dry + W * MAX_H;
      for (let y = 0; y < H; y++) {
        const dst = (y0 + y) * W, row = y * W;
        this.fb.set(this.layers.subarray(wet + row, wet + row + xi), dst);
        this.fb.set(this.layers.subarray(dry + row + xi, dry + row + W), dst + xi);
        if (xi < W && alpha > 0) this.fb[dst + xi] = blendLinear565(this.layers[dry + row + xi], this.layers[wet + row + xi], alpha);
      }
    }
    return this.fb;
  }
}

File: firmware/src/physical/render.cpp
#include "render.h"
#include "marks.h"
#include "output.h"
#include <string.h>
#ifdef ARDUINO
#include "esp_heap_caps.h"
#endif
namespace physical {
static constexpr int W=536,MAX_H=80;
static constexpr size_t LAYER_PIXELS=2*2*W*MAX_H;
static uint16_t *layers=nullptr;
struct Context { uint32_t gen=0; bool valid=false; };
static Context contexts[2];
static bool initAttempted=false;
bool init() {
  if(initAttempted)return layers!=nullptr;
  initAttempted=true;
#ifdef ARDUINO
  layers=(uint16_t*)heap_caps_malloc(LAYER_PIXELS*sizeof(uint16_t),MALLOC_CAP_SPIRAM);
#else
  static uint16_t hostLayers[LAYER_PIXELS]; // native reference: fixed BSS, never heap
  layers=hostLayers;
#endif
  return layers!=nullptr;
}
size_t memoryBytes(){return LAYER_PIXELS*sizeof(uint16_t)+sizeof(contexts);}
static RGB colour(const Params &p,const OpticalPath &path,float gain,float albedo) {
  return {p.exposure*(path.reflection+gain*albedo*beer(p.absorptionR,path.distance)),
          p.exposure*(path.reflection+gain*albedo*beer(p.absorptionG,path.distance)),
          p.exposure*(path.reflection+gain*albedo*beer(p.absorptionB,path.distance))};
}
static void rebuild(int tube,const Params &p) {
  int H=layout(p).H;float R=p.innerRadiusMm+p.wallThicknessMm,irradiance=backingLight(p);
  for(int medium=0;medium<2;medium++)for(int y=0;y<H;y++) {
    OpticalPath path=traceRow(((y+0.5f)/H*2-1)*R,medium==1,p);
    float gain=path.transmission*irradiance;
    uint16_t bg=__builtin_bswap16(quantize(colour(p,path,gain,p.backingReflectance)));
    uint16_t mark=__builtin_bswap16(quantize(colour(p,path,gain,p.marksReflectance)));
    float sy=(path.backY/R+1)*H*0.5f;
    uint16_t *row=layers+(tube*2+medium)*W*MAX_H+y*W;
    for(int x=0;x<W;x++)row[x]=path.transmission>0&&markAt(x,sy,H,tube)?mark:bg;
  }
}
void renderTube(int idx,float fill,const Params &p,uint32_t gen,uint16_t *strip) {
  if(!layers||!strip||idx<0||idx>1||!validate(p)||!isfinite(fill))return;
  Context &ctx=contexts[idx];
  if(!ctx.valid||ctx.gen!=gen){rebuild(idx,p);ctx.gen=gen;ctx.valid=true;}
  int H=layout(p).H;
  float edge=fmaxf(0,fminf(1,fill))*W;int xi=(int)floorf(edge);float alpha=edge-xi;
  const uint16_t *dry=layers+idx*2*W*MAX_H,*wet=dry+W*MAX_H;
  for(int y=0;y<H;y++) {
    int row=y*W;
    memcpy(strip+row,wet+row,xi*sizeof(uint16_t));
    memcpy(strip+row+xi,dry+row+xi,(W-xi)*sizeof(uint16_t));
    if(xi<W&&alpha>0)strip[row+xi]=__builtin_bswap16(blendLinear565(__builtin_bswap16(dry[row+xi]),__builtin_bswap16(wet[row+xi]),alpha));
  }
}
}
